### PR TITLE
NIFI-6061: Fix CLOB/BLOB handling in PutDatabaseRecord

### PR DESCRIPTION
**Pull request by mattyb149**
Co-authored-by: zhangcheng <zhangchengk@foxmail.com>

Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

Uses some code from @KuKuDeCheng #4551 to fix BLOB/CLOB handling in PutDatabaseRecord, without all the refactoring into database adapters. If the refactor is desired it can be done under an improvement Jira

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] Have you verified that the full build is successful on JDK 8?
- [ ] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java [hunk 1]
```diff
@@ -721,10 +722,30 @@ private void executeDML(final ProcessContext context, final ProcessSession sessi
...
+                                    if (sqlType == Types.BLOB || sqlType == Types.BINARY) {
+                                        if (currentValue instanceof Object[]) {
+                                            // Convert Object[] arrays to byte[]
+                                            Object[] src = (Object[]) currentValue;
+                                            byte[] dest = new byte[src.length];
+                                            for (int j = 0; j < src.length; j++) {
+                                                dest[j] = (byte) src[j];
+                                            }
+                                            currentValue = dest;
+                                        } else if (currentValue instanceof Byte[]) {
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java:734]: Based on the logic, will this block be exercised since the initial check handles `Object[]`?  Running the unit test in debug mode results in a `currentValue` of `Byte[]` hitting the above conditional and not this conditional.  At minimum, it seems like the ordering should be adjusted to check for `Byte[]` before `Object[]`.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java [hunk 2]
```diff
@@ -721,10 +722,30 @@ private void executeDML(final ProcessContext context, final ProcessSession sessi
                             try {
                                 DataType targetDataType = DataTypeUtils.getDataTypeFromSQLTypeValue(sqlType);
                                 if (targetDataType != null) {
-                                    currentValue = DataTypeUtils.convertType(
-                                            currentValue,
-                                            targetDataType,
-                                            fieldName);
+                                    if (sqlType == Types.BLOB || sqlType == Types.BINARY) {
+                                        if (currentValue instanceof Object[]) {
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java:726]: Although the conditional checks for `Object[]` the loop casts each element to a `byte`.  It seems like the check for `Object[]` should be more narrow, and perhaps an exception should be thrown if any other array object type is found?

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java [hunk 3]
```diff
@@ -776,6 +797,58 @@ private void executeDML(final ProcessContext context, final ProcessSession sessi
...
+        } else if (sqlType == Types.CLOB) {
+            if (value == null) {
+                try {
+                    ps.setNull(index, Types.CLOB);
+                } catch (SQLException e) {
+                    throw new IOException("Unable to setNull() on prepared statement", e);
+                }
+            } else {
+                try {
+                    ps.setString(index, value.toString());
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java:838]: Does handling a CLOB using setString() support all required use cases?

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java [hunk 4]
```diff
@@ -776,6 +797,58 @@ private void executeDML(final ProcessContext context, final ProcessSession sessi
...
+                    }
+                }
+                byte[] byteArray = (byte[]) value;
+                try (InputStream inputStream = new ByteArrayInputStream(byteArray)) {
+                    ps.setBlob(index, inputStream);
+                } catch (SQLException e) {
+                    throw new IOException("Unable to parse binary data " + value, e.getCause());
+                }
+            } else {
+                try (InputStream inputStream = new ByteArrayInputStream(value.toString().getBytes())) {
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java:823]: Should the specific character set be specified with the call to `getBytes()` or is it intentional to have this call be dependent on the runtime system properties?  Is it possible to add a unit test to exercise this block?

File: nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/util/DataTypeUtils.java [hunk 5]
```diff
@@ -1995,13 +1995,17 @@ public static DataType getDataTypeFromSQLTypeValue(final int sqlType) {
...
                 return RecordFieldType.STRING.getDataType();
             case Types.TIME:
                 return RecordFieldType.TIME.getDataType();
             case Types.TIMESTAMP:
                 return RecordFieldType.TIMESTAMP.getDataType();
             case Types.ARRAY:
                 return RecordFieldType.ARRAY.getDataType();
+            case Types.BINARY:
+            case Types.BLOB:
+                return RecordFieldType.ARRAY.getArrayDataType(RecordFieldType.BYTE.getDataType());
```
Comment (exceptionfactory) [inline nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/util/DataTypeUtils.java:2008]: It would be helpful to have a unit test for the changes to indicate the intended behavior.

**Review by exceptionfactory — COMMENTED** [review]
Thanks for putting together this PR @mattyb149.  There are a number of challenging edge cases, so see the comments for some questions on a few of the details.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java (hunk 2 above)
Comment (mattyb149) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java:726]: I waffled a bit on that point, but figured if the values could be cast to bytes we'd allow it. Then again, if the field is an Object[Integer], the user might think all their bytes are going in when in fact only 1/4 of them would. I will add an additional check for byte elements

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java (hunk 1 above)
Comment (mattyb149) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java:734]: Good point, will exchange the order

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java (hunk 4 above)
Comment (mattyb149) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java:823]: This is another one of those cases like the above, a corner case I don't even know exists. Maybe if the target column is a BLOB but the field is a String or something, basically any time the final "else" clause above is reached, where we just convert the data according to the field and column types without the special BLOB/CLOB handling. I can try some more unit tests to exercise this.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java (hunk 3 above)
Comment (mattyb149) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java:838]: That's one of the tough things about JDBC and RDBMSs, nobody plays by all the same rules. I tested setString() with PostgreSQL and Oracle and Derby (for the unit tests) and setString() works for CLOB, but setClob() isn't supported for Derby. I'll look at trying setClob() or setCharacterStream() first, and if it fails, fall back to trying setString()

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java (hunk 1 above)
Comment (mattyb149) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java:734]: Since Byte[] covers the actual use case, then due to the above comment I really should only have to check for null or byte[] since those will work downstream. Otherwise we don't know how to do the conversion so an exception should be thrown. Agreed?

Comment (exceptionfactory) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java:734]: That sounds good, it avoids potentially unexpected behavior, and if the scenario arises, implementing a new check for a more specific array object type seems like the way to go.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java (hunk 3 above)
Comment (exceptionfactory) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java:838]: Thanks for working through these details, very difficult when JDBC drivers do not behave in a predictable way.  Sounds like a good way forward.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java (hunk 4 above)
Comment (exceptionfactory) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java:826]: Thanks for making the updates, just noticed this line should be adjusted to explicitly specify UTF-8 for `getBytes()` as opposed to relying on platform defaults.

Comment (mattyb149) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java:826]: Oops missed that one, will update now

**Review by exceptionfactory — APPROVED** [review]
Thanks for working through the feedback and including additional tests @mattyb149! +1 Merging.
